Re-entrant navigation commits wrong DocumentLoader

roberto-apple opened the pull request:
#### 0803c0c0e2a6b07dfafe916e21f43bb9d685af41
<pre>
Re-entrant navigation commits wrong DocumentLoader
<a href="https://bugs.webkit.org/show_bug.cgi?id=302047">https://bugs.webkit.org/show_bug.cgi?id=302047</a>
<a href="https://rdar.apple.com/161084444">rdar://161084444</a>

Reviewed by NOBODY (OOPS!).

commitProvisionalLoad() captures which DocumentLoader to commit, but
transitionToCommitted() re-reads m_provisionalDocumentLoader instead of using
the captured value. When a child frame commits synchronously during the parent&apos;s
commit (e.g., &lt;embed&gt; with data: URL), the parser resumes and fires subsequent
handlers (e.g., &lt;svg onload&gt;) while parent commit is on the stack. If this
handler starts another navigation, BackForwardCache clears the original loader
and replaces it with the new one. transitionToCommitted() then reads the wrong
loader and commits it, orphaning the original document.

The fix passes the captured DocumentLoader as an explicit parameter to
transitionToCommitted(), ensuring the originally-intended loader is committed.
The re-entrancy check (provisionalLoader != m_provisionalDocumentLoader) now
aborts the commit when it detects the member variable changed.

The initial reported crash occurred in WebKitTestRunner when resetInternalsObject()
dereferenced the null frame pointer of the orphaned document during test cleanup. That
crash doesn&apos;t seem reproducible with run-webkit-tests because the harness terminates
before the bad commit completes.

The added layout test triggers the re-entrant scenario (style + embed + svg) and
navigates to a data: URL containing test completion logic. With the fix, the
first navigation is aborted when re-entrancy is detected, and the second
navigation proceeds cleanly, allowing testRunner.notifyDone() to execute.
Without the fix, the wrong loader is committed, the test document is orphaned,
and the test times out.

Test: loader/child-frame-navigation-during-commit.html

* LayoutTests/loader/child-frame-navigation-during-commit-expected.txt: Added.
* LayoutTests/loader/child-frame-navigation-during-commit.html: Added.
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::commitProvisionalLoad):
(WebCore::FrameLoader::transitionToCommitted):
* Source/WebCore/loader/FrameLoader.h:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0803c0c0e2a6b07dfafe916e21f43bb9d685af41

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133840 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6351 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/45030 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141417 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/85900 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/3c5785db-0b3b-47e1-8ab4-d081841c0eb2) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/135710 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6879 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/6215 "Built successfully") | [❌ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102396 "Found 1 new test failure: loader/child-frame-navigation-during-commit.html (failure)") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/69868 "Found 1 new test failure: loader/child-frame-navigation-during-commit.html (failure)") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/23a07eb5-db7a-4311-9ea9-32d043985e08) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136787 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/4850 "Found 1 new test failure: loader/child-frame-navigation-during-commit.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/119990 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83195 "Passed tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/b29332c9-982b-416e-b938-efcd5dc612ee) 
| | [❌ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4729 "Exiting early after 10 failures. 10 tests run. 1 failures") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2347 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/795 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113882 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38107 "Found 1 new test failure: loader/child-frame-navigation-during-commit.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144063 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6021 "Built successfully") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38685 "Found 1 new test failure: loader/child-frame-navigation-during-commit.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110756 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6103 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5136 "Found 1 new test failure: loader/child-frame-navigation-during-commit.html (failure)") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110954 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4585 "Passed tests") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116243 "Exiting early after 10 failures. 10 tests run. 1 failures") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/59768 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6073 "Built successfully") | [❌ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34526 "Found 1 new test failure: loader/child-frame-navigation-during-commit.html (failure)") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5918 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69537 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6164 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6027 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

RupinMittal commented on LayoutTests/TestExpectations:
```diff
@@ -2927,6 +2927,9 @@ loader/stateobjects/pushstate-size-iframe.html [ Slow ]
 loader/stateobjects/replacestate-size.html [ Slow ]
+    
+# Timing-dependent re-entrant navigation test. Multiple valid outputs expected.
+loader/child-frame-navigation-during-commit.html [ Pass Failure ]
```
I haven't taken a thorough look at the code change yet, but at first glance, this stands out. We don't want to add a test that could pass or fail. Is there a way to change this so the test is always passing?

roberto-apple replied:
Yeah, it differently isn't a good look, but more than that, it's not actually testing the crash in the way I previously thought so it will likely have to be removed, but I've been struggling a bit to find a working layout test (that crashes without fix, does not crash with fix, and has a deterministic output).

This was the original test case I've been working with:
```
<script>
    function load() {
        var dataURL = "data:,Hello%2C%20World%21";
        window.navigation.navigate(dataURL);
    }

    function load2() {
        var dataURL = "data:,Good%2C%20Bye!%21";
        window.navigation.navigate(dataURL);
    }
</script>

<body>
    <style onload="load()"></style>
    <embed src="data:text/html,foo">
    <svg onload="load2()"></svg>
</body>
```

I've only been able to reproduce the crash (on release) when calling that test case directly in WebKitTestRunner:
i.e., `DYLD_LIBRARY_PATH=$(pwd) DYLD_FRAMEWORK_PATH=$(pwd) ./WebKitTestRunner test-case.html`

but have been unable to do so when using `run-webkit-tests`. It looks like with run-webkit-tests script it doesn't appear to call the problematic code `WebCoreTestSupport::resetInternalsObject()`

roberto-apple committed:
Re-entrant navigation commits wrong DocumentLoader
https://bugs.webkit.org/show_bug.cgi?id=302047
rdar://161084444

Reviewed by NOBODY (OOPS!).

commitProvisionalLoad() captures which DocumentLoader to commit, but
transitionToCommitted() re-reads m_provisionalDocumentLoader instead of using
the captured value. When a child frame commits synchronously during the parent's
commit (e.g., <embed> with data: URL), the parser resumes and fires subsequent
handlers (e.g., <svg onload>) while parent commit is on the stack. If this
handler starts another navigation, BackForwardCache clears the original loader
and replaces it with the new one. transitionToCommitted() then reads the wrong
loader and commits it, orphaning the original document.

The fix passes the captured DocumentLoader as an explicit parameter to
transitionToCommitted(), ensuring the originally-intended loader is committed.
The re-entrancy check (provisionalLoader != m_provisionalDocumentLoader) now
aborts the commit when it detects the member variable changed.

The initial reported crash occurred in WebKitTestRunner when resetInternalsObject()
dereferenced the null frame pointer of the orphaned document during test cleanup. That
crash doesn't seem reproducible with run-webkit-tests because the harness terminates
before the bad commit completes.

The added layout test triggers the re-entrant scenario (style + embed + svg) and
navigates to a data: URL containing test completion logic. With the fix, the
first navigation is aborted when re-entrancy is detected, and the second
navigation proceeds cleanly, allowing testRunner.notifyDone() to execute.
Without the fix, the wrong loader is committed, the test document is orphaned,
and the test times out.

Test: loader/child-frame-navigation-during-commit.html

* LayoutTests/loader/child-frame-navigation-during-commit-expected.txt: Added.
* LayoutTests/loader/child-frame-navigation-during-commit.html: Added.
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::commitProvisionalLoad):
(WebCore::FrameLoader::transitionToCommitted):
* Source/WebCore/loader/FrameLoader.h:

basuke commented on LayoutTests/loader/child-frame-navigation-during-commit.html:
```diff
@@ -0,0 +1,32 @@
+        '  document.write("PASS: Successfully navigated to second navigation (GoodBye)");' +
+        '  testRunner.notifyDone();' +
+        '}' +
+        '<\/script>';
```
Use backtick to represent markup or script as string in the code these days.

basuke commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -2564,10 +2567,11 @@ void FrameLoader::transitionToCommitted(CachedPage* cachedPage)
     // current load or the two will stomp each other.
-    setDocumentLoader(m_provisionalDocumentLoader.copyRef());
-    if (originalProvisionalDocumentLoader != m_provisionalDocumentLoader)
+    // Use the captured provisionalLoader parameter, not m_provisionalDocumentLoader
```
I think this comment describes nothing. Your code is enough to describe the intention.

basuke commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -2531,14 +2531,18 @@ IGNORE_GCC_WARNINGS_END
 }
 
-void FrameLoader::transitionToCommitted(CachedPage* cachedPage)
+void FrameLoader::transitionToCommitted(DocumentLoader* provisionalLoader, CachedPage* cachedPage)
```
Should we accept `RefPtr<DocumentLoader>&&` instead of raw pointer like `setDocumentLoader()` does?

basuke replied:
I'm uncomfortable this pointer is kept as raw until after `setDocumentLoader` even if the caller keeps the reference for now. Code will change and each function site should follow the convention of Safer CPP.

basuke commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -2564,10 +2567,11 @@ void FrameLoader::transitionToCommitted(CachedPage* cachedPage)
-    setDocumentLoader(m_provisionalDocumentLoader.copyRef());
-    if (originalProvisionalDocumentLoader != m_provisionalDocumentLoader)
+    // Use the captured provisionalLoader parameter, not m_provisionalDocumentLoader
+    setDocumentLoader(RefPtr { provisionalLoader });
```
We can simply copy the reference here if we change it as `RefPtr<DocumentLoader>`.

basuke commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -2531,14 +2531,18 @@ IGNORE_GCC_WARNINGS_END
         return;
 
+    // Ensure we're committing the loader that was originally captured
+    ASSERT(provisionalLoader);
```
Reading `commitProvisionalLoad()`, it says `pdl` can be nullptr. See just after invocation of `transitionToCommitted()` above. If statement checks `pdl`.

basuke reviewed (changes requested):
I read your stack trace carefully and the whole synchronous call changes starts not in `transistionToCommitted()`, but `addIfCacheable()`. See stack trace #68. This means your change is also using wrong document loader to be passed to `transitionToCommitted()` because `pdl` is captured in the beginning of the function. I think this PR is off track.
